### PR TITLE
Github workflow that uses kitctl to bootstrap a test environment

### PR DESCRIPTION
**Pull request by Ganiredi**
Issue #, if available:

Description of changes:

This workflow automates the build, testing, and bootstrapping of a test environment on Kubernetes using KIT CLI and AWS credentials for validating changes made in the main branch using GitHub Actions.

For this workflow, initially created it using my own set of IAM role to run locally on my repo. However, in order to ensure compatibility and proper functionality with the awslabs/kubernetes-iteration-toolkit repository, it is necessary to update the the associated IAM role . 

### **Workflow in my forked repo:**

https://github.com/Ganiredi/kubernetes-iteration-toolkit/actions/runs/3992151267/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/Test-kitctl.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+          aws-region: us-west-2
+    - name: Builds the KIT CLI ,Provisions the Kit Environment and deletes it
+      run: |
+        GOOS=${{ matrix.os }} GOARCH=${{ matrix.arch }} go build -o ./bin/kitctl ./cmd/kitctl/
+        cd /home/runner/work/kubernetes-iteration-toolkit/kubernetes-iteration-toolkit/substrate/cmd/kitctl
+        go build -v .
+        RANDOM_SUFFIX=$(openssl rand -hex 5)
+        ./kitctl bootstrap kit-github-${RANDOM_SUFFIX}
+        export KUBECONFIG=/home/runner/.kit/env/kit-github-${RANDOM_SUFFIX}/etc/kubernetes/admin.conf
+        ./kitctl delete kit-github-${RANDOM_SUFFIX}
```
Comment (mengqiy) [inline .github/workflows/Test-kitctl.yml:46]: I'd suggest running kubectl command (e.g. kubectl version) here to do a simple sanity check.

File: .github/workflows/Test-kitctl.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          role-to-assume: arn:aws:iam::078365735624:role/github-Token-generate
+          role-session-name: samplerolesession
+          aws-region: us-west-2
+    - name: Builds the KIT CLI ,Provisions the Kit Environment and deletes it
+      run: |
+        GOOS=${{ matrix.os }} GOARCH=${{ matrix.arch }} go build -o ./bin/kitctl ./cmd/kitctl/
+        cd /home/runner/work/kubernetes-iteration-toolkit/kubernetes-iteration-toolkit/substrate/cmd/kitctl
+        go build -v .
+        RANDOM_SUFFIX=$(openssl rand -hex 5)
+        ./kitctl bootstrap kit-github-${RANDOM_SUFFIX}
```
Comment (mengqiy) [inline .github/workflows/Test-kitctl.yml:44]: We can use something provided by GitHub actions e.g. try `jobs.<job_id>`

File: .github/workflows/Test-kitctl.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      uses: aws-actions/configure-aws-credentials@v1
+      with:
+          role-to-assume: arn:aws:iam::078365735624:role/github-Token-generate
+          role-session-name: samplerolesession
+          aws-region: us-west-2
+    - name: Builds the KIT CLI ,Provisions the Kit Environment and deletes it
+      run: |
+        GOOS=${{ matrix.os }} GOARCH=${{ matrix.arch }} go build -o ./bin/kitctl ./cmd/kitctl/
+        cd /home/runner/work/kubernetes-iteration-toolkit/kubernetes-iteration-toolkit/substrate/cmd/kitctl
+        go build -v .
```
[the remark this answers is not in the document]
Comment (Ganiredi) [reply .github/workflows/Test-kitctl.yml:42]: Yes ,I will remove it in next commit.

File: .github/workflows/Test-kitctl.yml (hunk 1 above)
Comment (Ganiredi) [reply .github/workflows/Test-kitctl.yml:46]: I was thinking the same thing, I'll make sure to add that step in my next commit.

File: .github/workflows/depend.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+name: depend
+on:
```
Comment (mengqiy) [inline .github/workflows/depend.yml:2]: I believe we want to have pull request as a trigger.

Comment (Ganiredi) [reply .github/workflows/depend.yml:2]: Made changes ,Now the dependent part is not triggering.

**Comment by mengqiy** [comment]
After looking more closely at https://github.com/aws-actions/configure-aws-credentials, I guess we can work around the OIDC issue when run it for a pull request from a fork by using aws-access-key-id, aws-secret-access-key, and a role-to-assume. Can you please try that?

**Comment by hakuna-matatah** [comment]
I was wondering if you could expand the scope of this PR to spin up to test EKS Infra and tekton workflow ?

**Comment by Ganiredi** [comment]
> 

It was on my plate to do smoke test on guest Eks cluster as well as Guest Kit cluster.Have to scope the Tekton workflow.

**Comment by Ganiredi** [comment]
In order to use AWS resources in GitHub Action workflow there are some limitations ,When we a Create a PR from the fork to the upstream repository workflow wont works because GitHub Secrets are not passed to the runner when a workflow is triggered from a forked repository, In order to overcome to this @mengqiy and I came up with an workaround  to use a secondary GitHub Action Workflow that is triggered on [workflow_run](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run). Here first workflow action will be triggered and basing on that the second GA will authenticate properly(AWS creds). But there are certain limitations when we are triggering the workflow_run events ,Where the second workflow that was triggered by the first workflow will always run in the background and doesn't have visibility in `checks`.

References:
(1)[https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-featu[…]-repository/managing-github-actions-settings-for-a-repository](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories)

(2)https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow

**Pull request closed without merging**